File: iOSChatExample/iOSChatExample/Models/MessageContent.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import Foundation

// MARK: - MessageContent Protocol
protocol MessageContent {
    static func decode(from text: String) -> MessageContent?
}

// MARK: - Plain Text Content
struct PlainTextContent: MessageContent {
    let text: String

    static func decode(from text: String) -> MessageContent? {
        PlainTextContent(text: text)
    }
}

// MARK: - Generic Interactive Template
struct GenericInteractiveTemplate: Decodable {
    let templateType: String
    // Other properties common to all interactive message types, if any
}

// MARK: - Interactive Content Protocol
protocol InteractiveContent: MessageContent {
    static var templateType: String { get }
}

// MARK: - Quick Reply Content
struct QuickReplyElement: Codable {
    let title: String
}

struct QuickReplyContentData: Codable {
    let title: String
    let subtitle: String?
    let elements: [QuickReplyElement]
}

struct QuickReplyData: Codable {
    let content: QuickReplyContentData
}

struct QuickReplyTemplate: Codable {
    let templateType: String
    let version: String
    let data: QuickReplyData
}

struct QuickReplyContent: InteractiveContent {
    static let templateType = "QuickReply" // This should match the templateType value for Quick Replies in the JSON

    let title: String
    let subtitle: String?
    let options: [String]

    static func decode(from text: String) -> MessageContent? {
        guard let jsonData = text.data(using: .utf8) else { return nil }
        do {
            let quickReply = try JSONDecoder().decode(QuickReplyTemplate.self, from: jsonData)
            let options = quickReply.data.content.elements.map { $0.title }
            let title = quickReply.data.content.title
            let subtitle = quickReply.data.content.subtitle
            return QuickReplyContent(title: title, subtitle: subtitle, options: options)
        } catch {
            print("Error decoding QuickReplyContent: \(error)")
            return nil
        }
    }
}

// MARK: - List Picker Content

struct ListPickerElement: Codable, Hashable, Equatable {    
    let title: String
    let subtitle: String?
    let imageType: String?
    let imageData: String?
}

struct ListPickerContentData: Codable {
    let title: String
    let subtitle: String?
    let imageType: String?
    let imageData: String?
    let elements: [ListPickerElement]
}

struct ListPickerData: Codable {
    let content: ListPickerContentData
}

struct ListPickerTemplate: Codable {
    let templateType: String
    let version: String
    let data: ListPickerData
}

struct ListPickerContent: InteractiveContent {
    static let templateType = "ListPicker" // This should match the templateType value for List Pickers in the JSON

    let title: String
    let subtitle: String?
    let imageUrl: String?
    let options: [ListPickerElement]

    static func decode(from text: String) -> MessageContent? {
        guard let jsonData = text.data(using: .utf8) else { return nil }
        do {
            let listPicker = try JSONDecoder().decode(ListPickerTemplate.self, from: jsonData)
            let title = listPicker.data.content.title
            let subtitle = listPicker.data.content.subtitle
            let options = listPicker.data.content.elements
            let imageUrl = listPicker.data.content.imageData
            return ListPickerContent(title: title, subtitle: subtitle, imageUrl: imageUrl, options: options)
        } catch {
            print("Error decoding ListPickerContent: \(error)")
            return nil
        }
    }
}


// MARK: - Additional Interactive Content Types
// Add additional structs here following the pattern of QuickReplyContent for each new type of interactive content.

File: mobileChatExamples/androidChatExample/app/src/main/java/com/blitz/androidchatexample/models/MessageContent.kt
package com.blitz.androidchatexample.models

import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json

interface MessageContent {
}

// Plain Text
@Serializable
data class PlainTextContent(val text: String) : MessageContent {
    companion object{
        fun decode(text: String): MessageContent = PlainTextContent(text)
    }
}

// Generic Interactive Template
@Serializable
data class GenericInteractiveTemplate(
    val templateType: String
    // Other common properties
)

interface InteractiveContent : MessageContent {
}

// Quick Reply Content
@Serializable
data class QuickReplyElement(
    val title: String
)
@Serializable
data class QuickReplyContentData(
    val title: String,
    val subtitle: String? = null,
    val elements: List<QuickReplyElement>
)
@Serializable
data class QuickReplyData(
    val content: QuickReplyContentData
)
@Serializable
data class QuickReplyTemplate(
    val templateType: String,
    val version: String,
    val data: QuickReplyData
)
@Serializable
data class QuickReplyContent(
    val title: String,
    val subtitle: String? = null,
    val options: List<String>
) : InteractiveContent {
    companion object {
        const val templateType = "QuickReply"
        fun decode(text: String): InteractiveContent? {
            return try {
                val quickReply = Json.decodeFromString<QuickReplyTemplate>(text)
                val options = quickReply.data.content.elements.map { it.title }
                val title = quickReply.data.content.title
                val subtitle = quickReply.data.content.subtitle ?: "" // Fallback to empty string if null
                QuickReplyContent(title, subtitle, options)
            } catch (e: Exception) {
                println("Error decoding QuickReplyContent: ${e.message}")
                null
            }
        }
    }
}

// List Picker
@Serializable
data class ListPickerElement(
    val title: String,
    val subtitle: String? = null,
    val imageType: String? = null,
    val imageData: String? = null
)
@Serializable
data class ListPickerContentData(
    val title: String,
    val subtitle: String? = null,
    val imageType: String? = null,
    val imageData: String? = null,
    val elements: List<ListPickerElement>
)
@Serializable
data class ListPickerData(
    val content: ListPickerContentData
)
@Serializable
data class ListPickerTemplate(
    val templateType: String,
    val version: String,
    val data: ListPickerData
)
@Serializable
data class ListPickerContent(
    val title: String,
    val subtitle: String? = null,
    val imageUrl: String? = null,
    val options: List<ListPickerElement>
) : InteractiveContent {
    companion object {
        const val templateType = "ListPicker"
        fun decode(text: String): InteractiveContent? {
            return try {
                val listPicker = Json.decodeFromString<ListPickerTemplate>(text)
                val title = listPicker.data.content.title
                val subtitle = listPicker.data.content.subtitle ?: ""
                val options = listPicker.data.content.elements
                val imageUrl = listPicker.data.content.imageData ?: ""
                ListPickerContent(title, subtitle, imageUrl, options)
            } catch (e: Exception) {
                println("Error decoding ListPickerContent: ${e.message}")
                null
            }
        }
    }
}
